File: app/src/main/java/ru/iandreyshev/parserrss/models/rss/RssParserV2.java
package ru.iandreyshev.parserrss.models.rss;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.jdom2.Element;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import ru.iandreyshev.parserrss.models.repository.Article;
import ru.iandreyshev.parserrss.models.repository.Rss;

final class RssParserV2 extends RssParseEngine {
    private static final String FEED_NAME = "channel";
    private static final String FEED_TITLE = "title";
    private static final String FEED_ORIGIN = "link";
    private static final String FEED_DESCRIPTION = "description";

    private static final String ARTICLE_NAME = "item";
    private static final String ARTICLE_TITLE = "title";
    private static final String ARTICLE_ORIGIN = "link";
    private static final String ARTICLE_DESCRIPTION = "description";

    private static final String DATE_NAME = "pubDate";
    private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat(("EEE, d MMM yyyy HH:mm:ss Z"), Locale.ENGLISH);

    private static final String ARTICLE_IMG_NODE = "enclosure";
    private static final String ARTICLE_IMG_URL = "url";
    private static final String ARTICLE_IMG_TYPE = "type";

    @Override
    @Nullable
    protected Rss parseRss(final Element root) {
        final Element channel = root.getChild(FEED_NAME);
        final String title = channel.getChildText(FEED_TITLE);
        final String origin = channel.getChildText(FEED_ORIGIN);

        if (title == null || origin == null) {
            return null;
        }

        final Rss rss = new Rss(title, origin);
        rss.setDescription(channel.getChildText(FEED_DESCRIPTION));

        return rss;
    }

    @NonNull
    @Override
    protected ArrayList<Article> parseArticles(final Element root) {
        final Element channel = root.getChild(FEED_NAME);
        final List<Element> items = channel.getChildren(ARTICLE_NAME);
        final ArrayList<Article> result = new ArrayList<>();

        for (final Element item : items) {
            final Article article = parseArticle(item);

            if (article != null) {
                result.add(article);
            }
        }

        return result;
    }

    private Article parseArticle(final Element item) {
        final String title = item.getChildText(ARTICLE_TITLE);
        final String origin = item.getChildText(ARTICLE_ORIGIN);
        final String description = item.getChildText(ARTICLE_DESCRIPTION);

        if (title == null || origin == null || description == null) {
            return null;
        }

        final Article article = new Article(title, description, origin);
        parseArticleDate(item, article);
        parseArticleImage(item, article);

        return article;
    }

    private void parseArticleDate(final Element item, final Article article) {
        final String dateStr = item.getChildText(DATE_NAME);
        final Date date;

        try {
            date = DATE_FORMAT.parse(dateStr);
        } catch (Exception ex) {
            return;
        }

        article.setDate(date);
    }

    private void parseArticleImage(final Element item, final Article article) {
        final Element resource = item.getChild(ARTICLE_IMG_NODE);

        if (resource == null) {
            return;
        }

        final String url = resource.getAttributeValue(ARTICLE_IMG_URL);
        final String type = resource.getAttributeValue(ARTICLE_IMG_TYPE);

        if (url == null || type == null || !isValidImageType(type)) {
            return;
        }

        article.setImageUrl(url);
    }

    private boolean isValidImageType(String type) {
        return type.equals(MimeType.JPEG) || type.equals(MimeType.PNG);
    }
}

File: app/src/main/java/ru/iandreyshev/parserrss/models/rss/RssParserV2.kt
package ru.iandreyshev.parserrss.models.rss

import org.jdom2.Element

import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Date
import java.util.Locale

import ru.iandreyshev.parserrss.models.repository.Article
import ru.iandreyshev.parserrss.models.repository.Rss

internal class RssParserV2 : RssParseEngine() {

    companion object {
        private const val FEED_NAME = "channel"
        private const val FEED_TITLE = "title"
        private const val FEED_ORIGIN = "link"
        private const val FEED_DESCRIPTION = "description"

        private const val ARTICLE_NAME = "item"
        private const val ARTICLE_TITLE = "title"
        private const val ARTICLE_ORIGIN = "link"
        private const val ARTICLE_DESCRIPTION = "description"

        private const val DATE_NAME = "pubDate"
        private val DATE_FORMAT = SimpleDateFormat("EEE, d MMM yyyy HH:mm:ss Z", Locale.ENGLISH)

        private const val ARTICLE_IMG_NODE = "enclosure"
        private const val ARTICLE_IMG_URL = "url"
        private const val ARTICLE_IMG_TYPE = "type"

        private val IMAGE_TYPES = setOf(MimeType.JPEG, MimeType.PNG)
    }

    override fun parseRss(root: Element): Rss? {
        val channel = root.getChild(FEED_NAME)
        val title = channel.getChildText(FEED_TITLE)
        val origin = channel.getChildText(FEED_ORIGIN)

        if (title == null || origin == null) {
            return null
        }

        val rss = Rss(title, origin)
        rss.description = channel.getChildText(FEED_DESCRIPTION)

        return rss
    }

    override fun parseArticles(root: Element): ArrayList<Article> {
        val channel = root.getChild(FEED_NAME)
        val items = channel.getChildren(ARTICLE_NAME)
        val result = ArrayList<Article>()

        items.forEach { item ->
            val article = parseArticle(item)

            if (article != null) {
                result.add(article)
            }
        }

        return result
    }

    private fun parseArticle(item: Element): Article? {
        val title = item.getChildText(ARTICLE_TITLE)
        val origin = item.getChildText(ARTICLE_ORIGIN)
        val description = item.getChildText(ARTICLE_DESCRIPTION)

        if (title == null || origin == null || description == null) {
            return null
        }

        val article = Article(title, description, origin)
        parseArticleDate(item, article)
        parseArticleImage(item, article)

        return article
    }

    private fun parseArticleDate(item: Element, article: Article) {
        val dateStr = item.getChildText(DATE_NAME)
        val date: Date

        try {
            date = DATE_FORMAT.parse(dateStr)
        } catch (ex: Exception) {
            return
        }

        article.setDate(date)
    }

    private fun parseArticleImage(item: Element, article: Article) {
        val resource = item.getChild(ARTICLE_IMG_NODE) ?: return

        val url = resource.getAttributeValue(ARTICLE_IMG_URL)
        val type = resource.getAttributeValue(ARTICLE_IMG_TYPE)

        if (url == null || !IMAGE_TYPES.contains(type)) {
            return
        }

        article.imageUrl = url
    }
}
